xain_fl/config/schema.py

```diff
@@ -0,0 +1,358 @@
+    """
+    return And(int, lambda value: value > 0, error=error(key, expected_value))
+
+
+def positive_integer(key: str, expected_value: str = "a positive integer") -> Schema:
+    """Return a positive integer validator for the given configuration
+    item.
+
+    Args:
+
+        key: key of the configuration item
+        expected_value: description of the expected type of
+            value for this configuration item
+
+    """
+    return And(int, lambda value: value >= 0, error=error(key, expected_value))
+
+
+def url(key: str, expected_value: str = "a valid URL") -> Schema:
+    """Return a URL validator for the given configuration item.
+
+    Args:
+
+        key: key of the configuration item
+        expected_value: description of the expected type of
+            value for this configuration item
+
+    """
+
+    def is_valid_url(value):
+        try:
+            parsed = urllib.parse.urlparse(value)
+        except (ValueError, urllib.error.URLError):
+            return False
+        # A URL is considered valid if it has at least a scheme and a
+        # network location.
+        return all([parsed.scheme, parsed.netloc])
+
+    return And(str, is_valid_url, error=error(key, expected_value))
+
+
+def is_valid_hostname(value: str) -> bool:
+    """Return whether the given string is a valid hostname
+
+    Args:
+
+        value: string to check
+
+    Returns:
+
+        `True` if the given value is a valid hostname, `False`
+        otherwise
+    """
+    try:
+        idna.encode(value)
+    except idna.IDNAError:
+        return False
+    return True
+
+
+def is_valid_ip_address(value: str) -> bool:
+    """Return whether the given string is a valid IP address
+
+    Args:
+
+        value: string to check
+
+    Returns:
+
+        `True` if the given value is a valid IP address, `False`
+        otherwise
+    """
+    try:
+        ipaddress.ip_address(value)
+    except ipaddress.AddressValueError:
+        return False
+    return True
+
+
+def hostname_or_ip_address(
+    key: str, expected_value: str = "a valid domain name or IP address"
+) -> Schema:
+    """Return a hostname or IP address validator for the given
+    configuration item.
+
+    Args:
+
+        key: key of the configuration item
+        expected_value: description of the expected type of
+            value for this configuration item
+
+    """
+    return And(str, Or(is_valid_hostname, is_valid_ip_address), error=error(key, expected_value),)
+
+
+SERVER_SCHEMA = Schema(
+    {
+        Optional("host", default="localhost"): hostname_or_ip_address("server.host"),
+        Optional("port", default=50051): Use(
+            int, error=error("server.port", "a valid port number")
+        ),
+    }
+)
+
+AI_SCHEMA = Schema(
+    {
+        "initial_weights": Use(str, error=error("ai.initial_weights", "a valid path"),),
+        "rounds": strictly_positive_integer("ai.rounds"),
+        "epochs": strictly_positive_integer("ai.epochs"),
```

Comment:
Here we need to change the validation, so that we accept `0` as well.
See https://github.com/xainag/xain-fl/pull/227

Reply:
Thank you, done!